File: lib/data/vault.dart
import 'dart:convert';
import 'dart:io';
import 'dart:isolate';
import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';
import 'package:path_provider/path_provider.dart';

import 'models.dart';

const String kVaultFormat = 'ohmyssh.vault';
const int kVaultVersion = 1;
const String kVaultFileName = 'ohmyssh.vault';

const String _fieldFormat = 'format';
const String _fieldVersion = 'version';
const String _fieldKdf = 'kdf';
const String _fieldCipher = 'cipher';
const String _fieldNonce = 'nonce';
const String _fieldMac = 'mac';
const String _fieldData = 'data';

/// Written into the envelope, so raising it does not orphan existing vaults.
const int kKdfIterations = 120000;

class VaultException implements Exception {
  const VaultException(this.message);
  final String message;
  @override
  String toString() => message;
}

class WrongPasswordException extends VaultException {
  const WrongPasswordException() : super('Wrong master password');
}

/// Runs off the UI isolate: 120k HMAC rounds would otherwise drop frames.
Future<Uint8List> _deriveKey({
  required String password,
  required Uint8List salt,
  required int iterations,
}) {
  return Isolate.run(() async {
    final pbkdf2 = Pbkdf2(
      macAlgorithm: Hmac.sha256(),
      iterations: iterations,
      bits: 256,
    );
    final key = await pbkdf2.deriveKey(
      secretKey: SecretKey(utf8.encode(password)),
      nonce: salt,
    );
    return Uint8List.fromList(await key.extractBytes());
  });
}

Uint8List _randomBytes(int length) {
  final random = Random.secure();
  return Uint8List.fromList(
    List<int>.generate(length, (_) => random.nextInt(256)),
  );
}

class Vault {
  Vault._(this._file, this._key, this._salt, this._iterations);

  final File _file;
  Uint8List _key;
  Uint8List _salt;
  final int _iterations;

  File get file => _file;

  static final _aes = AesGcm.with256bits();

  static Future<File> defaultFile() async {
    final dir = await getApplicationSupportDirectory();
    await dir.create(recursive: true);
    return File('${dir.path}${Platform.pathSeparator}$kVaultFileName');
  }

  static Future<bool> exists() async => (await defaultFile()).exists();

  /// Creates a fresh vault, overwriting whatever is at [file].
  static Future<Vault> create({
    required String password,
    File? file,
    VaultData initial = const VaultData(),
  }) async {
    final target = file ?? await defaultFile();
    final salt = _randomBytes(16);
    final key = await _deriveKey(
      password: password,
      salt: salt,
      iterations: kKdfIterations,
    );
    final vault = Vault._(target, key, salt, kKdfIterations);
    await vault.save(initial);
    return vault;
  }

  static Future<Vault> open({required String password, File? file}) async {
    final target = file ?? await defaultFile();
    if (!await target.exists()) {
      throw const VaultException('Vault file not found');
    }

    final envelope = _readEnvelope(await target.readAsString());
    final kdf = envelope[_fieldKdf] as Map<String, dynamic>;
    final salt = base64Decode(kdf['salt'] as String);
    final iterations = (kdf['iterations'] as num).toInt();

    final key = await _deriveKey(
      password: password,
      salt: Uint8List.fromList(salt),
      iterations: iterations,
    );

    // Force a decrypt now so a bad password fails here and not on first read.
    await _decrypt(envelope, key);

    return Vault._(target, key, Uint8List.fromList(salt), iterations);
  }

  Future<VaultData> read() async {
    final envelope = _readEnvelope(await _file.readAsString());
    final clear = await _decrypt(envelope, _key);
    return VaultData.fromJson(jsonDecode(utf8.decode(clear)) as Map<String, dynamic>);
  }

  Future<void> save(VaultData data) async {
    final plain = utf8.encode(jsonEncode(data.toJson()));
    final nonce = _aes.newNonce();
    final box = await _aes.encrypt(
      plain,
      secretKey: SecretKey(_key),
      nonce: nonce,
    );

    final envelope = <String, dynamic>{
      _fieldFormat: kVaultFormat,
      _fieldVersion: kVaultVersion,
      _fieldKdf: {
        'algo': 'pbkdf2-hmac-sha256',
        'iterations': _iterations,
        'salt': base64Encode(_salt),
      },
      _fieldCipher: 'aes-256-gcm',
      _fieldNonce: base64Encode(nonce),
      _fieldMac: base64Encode(box.mac.bytes),
      _fieldData: base64Encode(box.cipherText),
    };

    // Write-then-rename so a crash mid-write cannot leave a truncated vault.
    final temp = File('${_file.path}.tmp');
    await temp.writeAsString(jsonEncode(envelope), flush: true);
    await temp.rename(_file.path);
  }

  /// Reads with the old key first, so this cannot reset a forgotten password.
  Future<void> changePassword(String newPassword) async {
    final data = await read();
    final salt = _randomBytes(16);
    _key = await _deriveKey(
      password: newPassword,
      salt: salt,
      iterations: _iterations,
    );
    _salt = salt;
    await save(data);
  }

  static Map<String, dynamic> _readEnvelope(String raw) {
    final Object? decoded;
    try {
      decoded = jsonDecode(raw);
    } catch (_) {
      throw const VaultException('Vault file is not valid JSON');
    }
    if (decoded is! Map<String, dynamic>) {
      throw const VaultException('Vault file has an unexpected shape');
    }
    if (decoded[_fieldFormat] != kVaultFormat) {
      throw const VaultException('Not an ohmyssh vault file');
    }
    final version = (decoded[_fieldVersion] as num?)?.toInt() ?? 0;
    if (version > kVaultVersion) {
      throw VaultException(
        'Vault was written by a newer version of the app (v$version)',
      );
    }
    return decoded;
  }

  static Future<List<int>> _decrypt(
    Map<String, dynamic> envelope,
    Uint8List key,
  ) async {
    try {
      return await _aes.decrypt(
        SecretBox(
          base64Decode(envelope[_fieldData] as String),
          nonce: base64Decode(envelope[_fieldNonce] as String),
          mac: Mac(base64Decode(envelope[_fieldMac] as String)),
        ),
        secretKey: SecretKey(key),
      );
    } on SecretBoxAuthenticationError {
      throw const WrongPasswordException();
    }
  }
}

File: shared/src/commonMain/kotlin/com/example/ohmyssh/data/Vault.kt
package com.example.ohmyssh.data

import com.example.ohmyssh.platform.AeadAuthenticationException
import com.example.ohmyssh.platform.AppFiles
import com.example.ohmyssh.platform.VaultCrypto
import com.example.ohmyssh.platform.readText
import com.example.ohmyssh.platform.writeTextAtomic
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlin.io.encoding.Base64

const val kVaultFormat = "ohmyssh.vault"
const val kVaultVersion = 1
const val kVaultFileName = "ohmyssh.vault"

private const val FIELD_FORMAT = "format"
private const val FIELD_VERSION = "version"
private const val FIELD_KDF = "kdf"
private const val FIELD_CIPHER = "cipher"
private const val FIELD_NONCE = "nonce"
private const val FIELD_MAC = "mac"
private const val FIELD_DATA = "data"

/** Written into the envelope, so raising it does not orphan existing vaults. */
const val kKdfIterations = 120000

open class VaultException(override val message: String) : Exception(message) {
    override fun toString(): String = message
}

class WrongPasswordException : VaultException("Wrong master password")

private val b64 = Base64.Default

private val json = Json { ignoreUnknownKeys = true }

class Vault private constructor(
    val path: String,
    private var key: ByteArray,
    private var salt: ByteArray,
    private val iterations: Int,
) {
    companion object {
        fun defaultPath(): String {
            val dir = AppFiles.appSupportDirectory()
            return "$dir${AppFiles.pathSeparator}$kVaultFileName"
        }

        fun exists(): Boolean = AppFiles.exists(defaultPath())

        suspend fun create(
            password: String,
            path: String = defaultPath(),
            initial: VaultData = VaultData(),
        ): Vault {
            val salt = VaultCrypto.randomBytes(16)
            val key = VaultCrypto.deriveKey(password, salt, kKdfIterations)
            val vault = Vault(path, key, salt, kKdfIterations)
            vault.save(initial)
            return vault
        }

        suspend fun open(password: String, path: String = defaultPath()): Vault {
            if (!AppFiles.exists(path)) throw VaultException("Vault file not found")
            return openFromText(AppFiles.readText(path), password, path)
        }

        suspend fun openFromText(raw: String, password: String, path: String): Vault {
            val envelope = readEnvelope(raw)
            val (key, salt, iterations) = deriveEnvelopeKey(envelope, password)

            // Force a decrypt now so a bad password fails here and not on first read.
            decrypt(envelope, key)

            return Vault(path, key, salt, iterations)
        }

        suspend fun decryptData(raw: String, password: String): VaultData {
            val envelope = readEnvelope(raw)
            val (key, _, _) = deriveEnvelopeKey(envelope, password)
            val clear = decrypt(envelope, key)
            return parsePayload(clear)
        }

        suspend fun decryptSidecar(raw: String, password: String, format: String): ByteArray {
            val envelope = readEnvelope(raw, format)
            val (key, _, _) = deriveEnvelopeKey(envelope, password)
            return decrypt(envelope, key)
        }

        internal fun parsePayload(clear: ByteArray): VaultData {
            val parsed = json.parseToJsonElement(clear.decodeToString())
            if (parsed !is JsonObject) throw VaultException("Vault payload has an unexpected shape")
            return VaultData.fromJson(parsed)
        }

        private suspend fun deriveEnvelopeKey(
            envelope: JsonObject,
            password: String,
        ): Triple<ByteArray, ByteArray, Int> {
            val kdf = envelope[FIELD_KDF] as? JsonObject
                ?: throw VaultException("Vault file has an unexpected shape")
            val salt = b64.decode(kdf.str("salt") ?: throw VaultException("Vault KDF salt missing"))
            val iterations = kdf.int("iterations") ?: kKdfIterations
            return Triple(VaultCrypto.deriveKey(password, salt, iterations), salt, iterations)
        }

        internal fun readEnvelope(raw: String, format: String = kVaultFormat): JsonObject {
            val decoded = try {
                json.parseToJsonElement(raw)
            } catch (_: Exception) {
                throw VaultException("Vault file is not valid JSON")
            }
            if (decoded !is JsonObject) throw VaultException("Vault file has an unexpected shape")
            if (decoded.str(FIELD_FORMAT) != format) {
                throw VaultException("Not an $format file")
            }
            val version = decoded.int(FIELD_VERSION) ?: 0
            if (version > kVaultVersion) {
                throw VaultException("Vault was written by a newer version of the app (v$version)")
            }
            return decoded
        }

        internal fun decrypt(envelope: JsonObject, key: ByteArray): ByteArray {
            try {
                return VaultCrypto.decrypt(
                    key = key,
                    nonce = b64.decode(envelope.str(FIELD_NONCE) ?: ""),
                    ciphertext = b64.decode(envelope.str(FIELD_DATA) ?: ""),
                    mac = b64.decode(envelope.str(FIELD_MAC) ?: ""),
                )
            } catch (_: AeadAuthenticationException) {
                throw WrongPasswordException()
            }
        }
    }

    fun readFileBytes(): ByteArray = AppFiles.readBytes(path)

    fun read(): VaultData {
        val envelope = readEnvelope(AppFiles.readText(path))
        return parsePayload(decrypt(envelope, key))
    }

    fun save(data: VaultData) {
        val plain = json.encodeToString(JsonObject.serializer(), data.toJson()).encodeToByteArray()
        // Write-then-rename so a crash mid-write cannot leave a truncated vault.
        AppFiles.writeTextAtomic(path, seal(kVaultFormat, plain))
    }

    private fun seal(format: String, clear: ByteArray): String {
        val nonce = VaultCrypto.randomBytes(12)
        val (cipherText, mac) = VaultCrypto.encrypt(key, nonce, clear)

        val envelope = buildJsonObject {
            put(FIELD_FORMAT, format)
            put(FIELD_VERSION, kVaultVersion)
            put(
                FIELD_KDF,
                buildJsonObject {
                    put("algo", "pbkdf2-hmac-sha256")
                    put("iterations", iterations)
                    put("salt", b64.encode(salt))
                },
            )
            put(FIELD_CIPHER, "aes-256-gcm")
            put(FIELD_NONCE, b64.encode(nonce))
            put(FIELD_MAC, b64.encode(mac))
            put(FIELD_DATA, b64.encode(cipherText))
        }
        return json.encodeToString(JsonObject.serializer(), envelope)
    }

    /**
     * State this port keeps beside the vault: same envelope, same key derived
     * from the same master password, its own file.
     *
     * It stays out of the vault itself deliberately. The Flutter app writes
     * `ohmyssh.vault` too, and would drop a key it knows nothing about on its
     * next save — so anything stored in there would survive only until the user
     * next opened the other app. Session history also grows without bound
     * relative to a credential list, and has no business riding along in a file
     * that gets exported and imported.
     */
    fun readSidecar(fileName: String, format: String): ByteArray? {
        val target = siblingPath(fileName)
        if (!AppFiles.exists(target)) return null
        return decrypt(readEnvelope(AppFiles.readText(target), format), key)
    }

    fun writeSidecar(fileName: String, format: String, clear: ByteArray) {
        AppFiles.writeTextAtomic(siblingPath(fileName), seal(format, clear))
    }

    fun deleteSidecar(fileName: String) {
        val target = siblingPath(fileName)
        if (AppFiles.exists(target)) AppFiles.delete(target)
    }

    fun readSidecarFileBytes(fileName: String): ByteArray? {
        val target = siblingPath(fileName)
        if (!AppFiles.exists(target)) return null
        return AppFiles.readBytes(target)
    }

    private fun siblingPath(fileName: String): String {
        val cut = path.lastIndexOf(AppFiles.pathSeparator)
        if (cut < 0) return fileName
        return path.substring(0, cut + AppFiles.pathSeparator.length) + fileName
    }

    /**
     * Reads with the old key first, so this cannot reset a forgotten password.
     * [sidecars] are name-to-format pairs read out under the old key and written
     * back under the new one, since nothing else would ever be able to open
     * them again.
     */
    suspend fun changePassword(newPassword: String, sidecars: List<Pair<String, String>> = emptyList()) {
        val data = read()
        val carried = sidecars.mapNotNull { (fileName, format) ->
            val clear = runCatching { readSidecar(fileName, format) }.getOrNull() ?: return@mapNotNull null
            Triple(fileName, format, clear)
        }

        val newSalt = VaultCrypto.randomBytes(16)
        key = VaultCrypto.deriveKey(newPassword, newSalt, iterations)
        salt = newSalt

        save(data)
        for ((fileName, format, clear) in carried) writeSidecar(fileName, format, clear)
    }
}
